### PR TITLE
1418 chargeowner match sender validation v2

### PR DESCRIPTION
**Pull request by jonasdmoeller**
## Description

Added document validation rule to ensure charge owners of all operations match sender id of market participant sending the message.

### CODE REVIEW

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargePriceEventHandler.cs [hunk 1]
```diff
@@ -113,7 +113,7 @@ public async Task HandleAsync(ChargePriceCommandReceivedEvent commandReceivedEve
             // With story 1411 below log entry will be replaced with 'await _unitOfWork.SaveChangesAsync().ConfigureAwait(false)';
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargePriceEventHandler.cs:113]:
```suggestion
            // TODO: With story 1411 below log entry will be replaced with 'await _unitOfWork.SaveChangesAsync().ConfigureAwait(false)';
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargePriceEventHandlerDeprecated.cs [hunk 2]
```diff
@@ -73,7 +73,7 @@ public async Task HandleAsync(ChargeCommandReceivedEvent commandReceivedEvent)
                 var charge = await GetChargeAsync(operation).ConfigureAwait(false);
                 if (charge is null)
                 {
-                    throw new InvalidOperationException($"Charge ID '{operation.ChargeId}' does not exist.");
+                    throw new InvalidOperationException($"Charge ID '{operation.SenderProvidedChargeId}' does not exist.");
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargePriceEventHandlerDeprecated.cs:76]: Is it possible to add a test for this failure scenario?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargePriceEventHandler.cs [hunk 3]
```diff
@@ -83,14 +83,14 @@ public async Task HandleAsync(ChargePriceCommandReceivedEvent commandReceivedEve
                 {
                     if (charge is null)
                     {
-                        throw new InvalidOperationException($"Charge ID '{operation.ChargeId}' does not exist.");
+                        throw new InvalidOperationException($"Charge ID '{operation.SenderProvidedChargeId}' does not exist.");
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargePriceEventHandler.cs:86]: Is it possible to add a test for this failure scenario?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Acknowledgement/ChargePricesUpdatedPublisher.cs [hunk 4]
```diff
@@ -32,9 +32,9 @@ public class ChargePricesUpdatedPublisher : IChargePricesUpdatedPublisher
             _chargePricesUpdatedEventFactory = chargePricesUpdatedEventFactory;
         }
 
-        public async Task PublishChargePricesAsync(ChargeOperationDto chargeOperationDto)
+        public async Task PublishChargePricesAsync(ChargeInformationOperationDto chargeInformationOperationDto)
         {
-            var prices = _chargePricesUpdatedEventFactory.Create(chargeOperationDto);
+            var prices = _chargePricesUpdatedEventFactory.Create(chargeInformationOperationDto);
             await _messagePricesDispatcher.DispatchAsync(prices).ConfigureAwait(false);
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Acknowledgement/ChargePricesUpdatedPublisher.cs:38]: Unrelatest to this story, but we should add a `ChargePricesUpdatedPublisherTests` class

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Services/ChargePriceRejectionService.cs [hunk 5]
```diff
@@ -38,7 +38,9 @@ public ChargePriceRejectionService(ILoggerFactory loggerFactory)
             foreach (var chargePriceOperationDto in rejectedPriceOperations)
             {
                 _logger.LogInformation(
-                    $"{chargePriceOperationDto.ChargeId} rejected price operations was persisted. With errors: {PrintInvalidRules(documentValidationResult)}");
+                    "{SenderProvidedChargeId} rejected price operations was persisted. With errors: {Errors}",
+                    chargePriceOperationDto.SenderProvidedChargeId,
+                    PrintInvalidRules(documentValidationResult));
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Services/ChargePriceRejectionService.cs:43]: Let's add tests for the logging in the two `SaveRejectionsAsync` methods

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeInformationCommands/Validation/InputValidation/ValidationRules/ChargeOwnerMustMatchSenderRule.cs [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+namespace GreenEnergyHub.Charges.Domain.Dtos.ChargeInformationCommands.Validation.InputValidation.ValidationRules
+{
+    public class ChargeOwnerMustMatchSenderRule : IValidationRule
+    {
+        private const int MinimumChargeOwnerLength = 13;
+        private const int MaximumChargeOwnerLength = 16;
+        private readonly ChargeInformationOperationDto _chargeInformationOperationDto;
+
+        public ChargeOwnerMustMatchSenderRule(ChargeInformationOperationDto chargeInformationOperationDto)
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeInformationCommands/Validation/InputValidation/ValidationRules/ChargeOwnerMustMatchSenderRule.cs:25]: Let's add `ChargeOwnerMustMatchSenderRuleTests`

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeInformationCommands/Validation/InputValidation/ValidationRules/ChargeOwnerMustMatchSenderRule.cs [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+        {
+            _chargeInformationOperationDto = chargeInformationOperationDto;
+        }
+
+        public ValidationRuleIdentifier ValidationRuleIdentifier => ValidationRuleIdentifier.ChargeOwnerHasLengthLimits;
+
+        public bool IsValid =>
+            _chargeInformationOperationDto.ChargeOwner.Length is >= MinimumChargeOwnerLength and <= MaximumChargeOwnerLength;
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeInformationCommands/Validation/InputValidation/ValidationRules/ChargeOwnerMustMatchSenderRule.cs:33]: How does the `Lenght` property verify that `ChargeOwner` matches `Sender`?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/Validation/DocumentValidation/ValidationRules/SenderMustMatchChargeOwnersRule.cs [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+        private readonly IEnumerable<string> _chargeOwners;
+
+        public SenderMustMatchChargeOwnersRule(DocumentDto documentDto, IEnumerable<string> chargeOwners)
+        {
+            _documentDto = documentDto;
+            _chargeOwners = chargeOwners;
+        }
+
+        public ValidationRuleIdentifier ValidationRuleIdentifier =>
+            ValidationRuleIdentifier.SenderMustMatchChargeOwners;
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/Validation/DocumentValidation/ValidationRules/SenderMustMatchChargeOwnersRule.cs:33]: Let's add a test for the ValidationRuleIdentifier in `SenderMustMatchChargeOwnersRuleTests`

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/Validation/ValidationRuleIdentifier.cs [hunk 9]
```diff
@@ -61,5 +61,6 @@ public enum ValidationRuleIdentifier
         TaxIndicatorMustBeFalseForSubscription = 47, // VR920 / D14
         TaxIndicatorMustBeFalseForFee = 48, // VR921 / D14
         ChargeOperationIdLengthValidation = 49, // VR922 / E86
+        SenderMustMatchChargeOwners = 50, // TODO: VR??? / ??
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/Validation/ValidationRuleIdentifier.cs:64]: Which VR?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure.Core/Cim/ReasonCode.cs [hunk 10]
```diff
@@ -111,5 +111,6 @@ public enum ReasonCode
         E91 = 191,
         E97 = 197,
         E98 = 198,
+        E99 = 999, // TODO: Find correct reason code
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure.Core/Cim/ReasonCode.cs:114]: Remember this one, before merging to main.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.MessageHub/Models/Shared/CimValidationErrorCodeFactory.cs [hunk 11]
```diff
@@ -69,6 +69,7 @@ public ReasonCode Create(ValidationRuleIdentifier validationRuleIdentifier)
                 ValidationRuleIdentifier.TaxIndicatorMustBeFalseForFee => ReasonCode.D14,
                 ValidationRuleIdentifier.TaxIndicatorMustBeFalseForSubscription => ReasonCode.D14,
                 ValidationRuleIdentifier.ChargeOperationIdLengthValidation => ReasonCode.E86,
+                ValidationRuleIdentifier.SenderMustMatchChargeOwners => ReasonCode.E99,
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.MessageHub/Models/Shared/CimValidationErrorCodeFactory.cs:72]: Remember to change this, before merging to main.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/DocumentValidation/ValidationRules/SenderMustMatchChargeOwnersRuleTest.cs [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+            string chargeOwner,
+            bool isValid)
+        {
+            var marketParticipant = new MarketParticipantDtoBuilder().WithId(senderId).Build();
+            var document = new DocumentDtoBuilder().WithSender(marketParticipant).Build();
+            var owners = new[] { chargeOwner, chargeOwner, chargeOwner };
+            var sut = new SenderMustMatchChargeOwnersRule(document, owners);
+            sut.IsValid.Should().Be(isValid);
+        }
+    }
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/DocumentValidation/ValidationRules/SenderMustMatchChargeOwnersRuleTest.cs:40]: Not sure if it makes sense, but should we add another test where not all owners are the same?
`var owners = new[] { chargeOwner, chargeOwner, someOtherChargeOwner };`

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargePriceEventHandler.cs (hunk 1 above)
Comment (jonasdmoeller) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargePriceEventHandler.cs:113]: I dont like having TODO comments in main. This should be reflected in a story related to the change instead.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargePriceEventHandlerDeprecated.cs (hunk 2 above)
Comment (jonasdmoeller) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargePriceEventHandlerDeprecated.cs:76]: This is just a property rename. If further tests are needed for a failure scenario related to the price event handling, then lets add a story for that.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargePriceEventHandler.cs (hunk 3 above)
Comment (jonasdmoeller) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargePriceEventHandler.cs:86]: Same as above

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Acknowledgement/ChargePricesUpdatedPublisher.cs (hunk 4 above)
Comment (jonasdmoeller) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Acknowledgement/ChargePricesUpdatedPublisher.cs:38]: I agree. Lets add a story for that.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Services/ChargePriceRejectionService.cs (hunk 5 above)
Comment (jonasdmoeller) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Services/ChargePriceRejectionService.cs:43]: Same as above.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeInformationCommands/Validation/InputValidation/ValidationRules/ChargeOwnerMustMatchSenderRule.cs (hunk 6 above)
Comment (jonasdmoeller) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeInformationCommands/Validation/InputValidation/ValidationRules/ChargeOwnerMustMatchSenderRule.cs:25]: This file is actually not intended to be committed :) The rule is implemented in `SenderMustMatchChargeOwnersRule` that has corresponding `SenderMustMatchChargeOwnersRuleTest`

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeInformationCommands/Validation/InputValidation/ValidationRules/ChargeOwnerMustMatchSenderRule.cs (hunk 7 above)
Comment (jonasdmoeller) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeInformationCommands/Validation/InputValidation/ValidationRules/ChargeOwnerMustMatchSenderRule.cs]: Same as above

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/Validation/DocumentValidation/ValidationRules/SenderMustMatchChargeOwnersRule.cs (hunk 8 above)
Comment (jonasdmoeller) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/Validation/DocumentValidation/ValidationRules/SenderMustMatchChargeOwnersRule.cs:33]: Done

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/DocumentValidation/ValidationRules/SenderMustMatchChargeOwnersRuleTest.cs (hunk 12 above)
Comment (jonasdmoeller) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/DocumentValidation/ValidationRules/SenderMustMatchChargeOwnersRuleTest.cs:40]: Done

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure.Core/Cim/ReasonCode.cs (hunk 10 above)
Comment (jonasdmoeller) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure.Core/Cim/ReasonCode.cs]: Fixed

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.MessageHub/Models/Shared/CimValidationErrorCodeFactory.cs (hunk 11 above)
Comment (jonasdmoeller) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.MessageHub/Models/Shared/CimValidationErrorCodeFactory.cs]: Fixed

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure.Core/Cim/ValidationErrors/CimValidationErrorTextTemplateMessages.cs [hunk 13]
```diff
@@ -37,6 +37,10 @@ public static class CimValidationErrorTextTemplateMessages
         public const string SenderIsMandatoryTypeValidationErrorText =
             "Sender is missing for message {{DocumentId}}.";
 
+        [ErrorMessageFor(ValidationRuleIdentifier.ChargeOwnerMustMatchSender)]
+        public const string ChargeOwnerMustMatchSenderErrorText =
+            "The specified charge type owner {{ChargeOwner}} do not match sender {{DocumentSenderId}} and is therefore not authorized to change charge id {{DocumentSenderProvidedChargeId}} of type {{ChargeType}}";
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure.Core/Cim/ValidationErrors/CimValidationErrorTextTemplateMessages.cs:42]:
```suggestion
            "The specified charge owner {{ChargeOwner}} do not match sender {{DocumentSenderId}} and is therefore not authorized to change charge id {{DocumentSenderProvidedChargeId}} of type {{ChargeType}}";
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/DocumentValidation/ValidationRules/ChargeOwnerMustMatchSenderRuleTests.cs [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+// limitations under the License.
+
+using FluentAssertions;
+using GreenEnergyHub.Charges.Domain.Dtos.Validation;
+using GreenEnergyHub.Charges.Domain.Dtos.Validation.DocumentValidation.ValidationRules;
+using GreenEnergyHub.Charges.Tests.Builders.Command;
+using Xunit;
+using Xunit.Categories;
+
+namespace GreenEnergyHub.Charges.Tests.Domain.Dtos.DocumentValidation.ValidationRules
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/DocumentValidation/ValidationRules/ChargeOwnerMustMatchSenderRuleTests.cs]: Move to namespace equivalent to `ChargeOwnerMustMatchSenderRule`

**Review by HenrikSommer — APPROVED** [review]

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/DocumentValidation/ValidationRules/ChargeOwnerMustMatchSenderRuleTests.cs (hunk 14 above)
Comment (jonasdmoeller) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/DocumentValidation/ValidationRules/ChargeOwnerMustMatchSenderRuleTests.cs]: Done

**Pull request merged**